File: src/native/apple_speech.cc
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Simen. All rights reserved.
 *  Licensed under the MIT License.
 *--------------------------------------------------------------------------------------------*/

#include <napi.h>
#include <string>
#include "apple_speech.h"

#if defined(__APPLE__)

static Napi::Value TranscribeFile(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();

    if (info.Length() < 1 || !info[0].IsString()) {
        Napi::TypeError::New(env, "File path string expected").ThrowAsJavaScriptException();
        return env.Null();
    }

    std::string filePath = info[0].As<Napi::String>().Utf8Value();
    std::string locale = "zh-CN";
    double timeout = 60.0;

    if (info.Length() >= 2 && info[1].IsString()) {
        locale = info[1].As<Napi::String>().Utf8Value();
    }

    if (info.Length() >= 3 && info[2].IsNumber()) {
        timeout = info[2].As<Napi::Number>().DoubleValue();
    }

    char* result = appleSpeechTranscribeFile(filePath.c_str(), locale.c_str(), timeout);
    if (!result) {
        return env.Null();
    }

    std::string json(result);
    appleSpeechFreeString(result);

    return Napi::String::New(env, json);
}

static Napi::Value IsAvailable(const Napi::CallbackInfo& info) {
    return Napi::Boolean::New(info.Env(), appleSpeechIsAvailable());
}

static Napi::Object Init(Napi::Env env, Napi::Object exports) {
    exports.Set("transcribeFile", Napi::Function::New(env, TranscribeFile));
    exports.Set("isAvailable", Napi::Function::New(env, IsAvailable));
    return exports;
}

NODE_API_MODULE(simen_speech_recognizer, Init)

#else

// Stub for non-macOS platforms
static Napi::Object Init(Napi::Env env, Napi::Object exports) {
    return exports;
}

NODE_API_MODULE(simen_speech_recognizer, Init)

#endif

File: src/native/AppleSpeech.swift
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Simen. All rights reserved.
 *  Licensed under the MIT License.
 *--------------------------------------------------------------------------------------------*/

import Foundation
import Speech

// MARK: - Result Structure

struct TranscriptionResult: Codable {
    let success: Bool
    let text: String
    let error: String?
}

// MARK: - Internal Helpers

private func encodeResult(_ result: TranscriptionResult) -> String {
    let encoder = JSONEncoder()
    guard let data = try? encoder.encode(result),
          let json = String(data: data, encoding: .utf8) else {
        return "{\"success\":false,\"text\":\"\",\"error\":\"JSON encoding failed\"}"
    }
    return json
}

private func errorResult(_ message: String) -> String {
    return encodeResult(TranscriptionResult(success: false, text: "", error: message))
}

private func successResult(_ text: String) -> String {
    return encodeResult(TranscriptionResult(success: true, text: text, error: nil))
}

// MARK: - Synchronous Transcription

private func transcribeSync(fileURL: URL, locale: String, timeoutSeconds: Double) -> String {
    guard let speechRecognizer = SFSpeechRecognizer(locale: Locale(identifier: locale)) else {
        return errorResult("Failed to create speech recognizer for locale: \(locale)")
    }

    guard speechRecognizer.isAvailable else {
        return errorResult("Speech recognizer is not available")
    }

    let request = SFSpeechURLRecognitionRequest(url: fileURL)
    request.shouldReportPartialResults = false
    request.taskHint = .dictation
    request.requiresOnDeviceRecognition = false

    let semaphore = DispatchSemaphore(value: 0)
    var resultText: String = ""
    var resultError: String? = nil

    speechRecognizer.recognitionTask(with: request) { result, error in
        if let error = error {
            resultError = error.localizedDescription
            semaphore.signal()
            return
        }
        if let result = result, result.isFinal {
            resultText = result.bestTranscription.formattedString
            semaphore.signal()
        }
    }

    let timeout = DispatchTime.now() + timeoutSeconds
    if semaphore.wait(timeout: timeout) == .timedOut {
        return errorResult("Transcription timed out")
    }

    if let error = resultError {
        return errorResult(error)
    }

    return successResult(resultText)
}

// MARK: - C Interface

@_cdecl("appleSpeechTranscribeFile")
public func appleSpeechTranscribeFile(
    _ filePath: UnsafePointer<CChar>?,
    _ locale: UnsafePointer<CChar>?,
    _ timeoutSeconds: Double
) -> UnsafeMutablePointer<CChar>? {
    guard let filePath = filePath else {
        return strdup(errorResult("File path is nil"))
    }

    let path = String(cString: filePath)
    let loc = locale != nil ? String(cString: locale!) : "zh-CN"
    let timeout = timeoutSeconds > 0 ? timeoutSeconds : 60.0

    let fileURL = URL(fileURLWithPath: path)

    guard FileManager.default.fileExists(atPath: path) else {
        return strdup(errorResult("File not found: \(path)"))
    }

    let result = transcribeSync(fileURL: fileURL, locale: loc, timeoutSeconds: timeout)
    return strdup(result)
}

@_cdecl("appleSpeechFreeString")
public func appleSpeechFreeString(_ ptr: UnsafeMutablePointer<CChar>?) {
    if let ptr = ptr {
        free(ptr)
    }
}

@_cdecl("appleSpeechIsAvailable")
public func appleSpeechIsAvailable() -> Bool {
    guard let recognizer = SFSpeechRecognizer() else {
        return false
    }
    return recognizer.isAvailable
}
